flink-core/src/main/java/org/apache/flink/util/JavaGcCleanerWrapper.java

```diff
@@ -0,0 +1,220 @@
+			} catch (NoSuchMethodException e) {
+				throw new FlinkRuntimeException("Failed to find Java legacy Cleaner#clean method", e);
+			}
+			return new LegacyCleanerFactory(cleanerCreateMethod, cleanerCleanMethod);
+		}
+	}
+
+	private static class LegacyCleanerFactory implements CleanerFactory {
+		private final Method cleanerCreateMethod;
+		private final Method cleanerCleanMethod;
+
+		private LegacyCleanerFactory(Method cleanerCreateMethod, Method cleanerCleanMethod) {
+			this.cleanerCreateMethod = cleanerCreateMethod;
+			this.cleanerCleanMethod = cleanerCleanMethod;
+		}
+
+		@Override
+		public Runnable create(Object owner, Runnable cleanOperation) {
+			Object cleaner;
+			try {
+				cleaner = cleanerCreateMethod.invoke(null, owner, cleanOperation);
+			} catch (IllegalAccessException | InvocationTargetException e) {
+				throw new FlinkRuntimeException("Failed to create a Java legacy Cleaner", e);
+			}
+			String ownerString = owner.toString(); // lambda should not capture the owner object
+			return () -> {
+				try {
+					cleanerCleanMethod.invoke(cleaner);
+				} catch (IllegalAccessException | InvocationTargetException e) {
+					String message = String.format("FATAL UNEXPECTED - Failed to invoke a Java legacy Cleaner for %s", ownerString);
+					LOG.error(message, e);
+					throw new FlinkRuntimeException(message, e);
+				}
+			};
+		}
+	}
+
+	private enum Java9CleanerProvider implements CleanerProvider {
+		INSTANCE;
+
+		@Override
+		public String getCleanerClassName() {
+			return JAVA9_CLEANER_CLASS_NAME;
+		}
+
+		@Override
+		public CleanerFactory createCleanerFactory(Class<?> cleanerClass) {
+			Method cleanerCreateMethod;
+			try {
+				cleanerCreateMethod = cleanerClass.getMethod("create");
+			} catch (NoSuchMethodException e) {
+				throw new FlinkRuntimeException("Failed to find Java 9 Cleaner#create method", e);
+			}
+			Object cleaner;
+			try {
+				cleaner = cleanerCreateMethod.invoke(null);
+			} catch (IllegalAccessException | InvocationTargetException e) {
+				throw new FlinkRuntimeException("Failed to create a Java 9 Cleaner", e);
+			}
+			Method cleanerRegisterMethod;
+			try {
+				cleanerRegisterMethod = cleanerClass.getMethod("register", Object.class, Runnable.class);
+			} catch (NoSuchMethodException e) {
+				throw new FlinkRuntimeException("Failed to find Java 9 Cleaner#create method", e);
+			}
+			Class<?> cleanableClass;
+			try {
+				cleanableClass = Class.forName("java.lang.ref.Cleaner$Cleanable");
+			} catch (ClassNotFoundException e) {
+				throw new FlinkRuntimeException("Failed to find Java 9 Cleaner#Cleanable class", e);
+			}
+			Method cleanMethod;
+			try {
+				cleanMethod = cleanableClass.getMethod("clean");
+			} catch (NoSuchMethodException e) {
+				throw new FlinkRuntimeException("Failed to find Java 9 Cleaner$Cleanable#clean method", e);
+			}
+			return new Java9CleanerFactory(cleaner, cleanerRegisterMethod, cleanMethod);
+		}
```

Comment:
This method could be a bit cleaner if we factor out the individual steps into methods which are responsible for the error handling.